File: 20-actors/shionome/methods/test_social.py
"""test_social.py — 潮目 (shionome) dry-run social post + no-trade body scan. ADR-2606072200."""
from __future__ import annotations

import social
from _t import expect_raises, run

_SRC = ["https://fred.stlouisfed.org/", "https://www.ici.org/research"]
_NET = [{"bucket": "us-equities", "label": "US equities", "net": 13.1, "inflow": 23.2, "outflow": 10.1},
        {"bucket": "us-govt-bonds", "label": "US Treasuries", "net": -21.7, "inflow": 0.0, "outflow": 21.7}]
_ROT = [{"from": "us-govt-bonds", "from_label": "US Treasuries", "to": "us-equities",
         "to_label": "US equities", "magnitude": 12.4}]
_REGIME = {"regime": "risk-on", "risk_net": 27.3, "safe_net": -26.6, "no_trade_notice": True}


def test_netflow_post_pins_invariants():
    p = social.draft_netflow_post(_NET, _SRC)
    assert p[":post/status"] == ":dry-run"
    assert p[":post/is-mirror"] is True
    assert p[":post/no-trade-notice"] is True
    assert p[":post/server-held-key"] is False


def test_netflow_post_body_has_disclaimer():
    p = social.draft_netflow_post(_NET, _SRC)
    assert "トレードはしない" in p[":post/body"]


def test_rotation_post_ok():
    p = social.draft_rotation_post(_ROT, _SRC)
    assert "US Treasuries" in p[":post/body"] and p[":post/status"] == ":dry-run"


def test_regime_post_states_descriptor():
    p = social.draft_regime_post(_REGIME, _SRC)
    assert "risk-on" in p[":post/body"]
    assert "助言ではありません" in p[":post/body"]


def test_post_requires_two_sources_g3():
    expect_raises(lambda: social.draft_netflow_post(_NET, ["only-one"]), contains="G3")


def test_post_refuses_denied_source():
    expect_raises(lambda: social.draft_netflow_post(_NET, ["bloomberg terminal", "x"]), contains="Rider")


def test_no_trade_guard_blocks_buy():
    expect_raises(lambda: social._guard_no_trade("you should buy equities"), contains="G2")


def test_no_trade_guard_blocks_japanese():
    expect_raises(lambda: social._guard_no_trade("目標株価は5000円"), contains="G2")


def test_no_trade_guard_allows_disclaimer():
    # the disclaimer NAMES the prohibited acts; it must not trip the guard
    social._guard_no_trade(social.DISCLAIMER)


def test_build_live_refuses_g8():
    expect_raises(lambda: social.build_live(), contains="G8")


def test_empty_rotation_post_safe():
    p = social.draft_rotation_post([], _SRC)
    assert p[":post/status"] == ":dry-run"


if __name__ == "__main__":
    run("social", [(n, f) for n, f in sorted(globals().items())
                   if n.startswith("test_") and callable(f)])

File: 20-actors/shionome/methods/test_social.clj
#!/usr/bin/env bb
;; test_social.clj — 潮目 (shionome) dry-run social post + no-trade body scan. ADR-2606072200.
(ns shionome.methods.test-social
  (:require [shionome.methods.social :as social]
            [clojure.test :refer [deftest is testing run-tests]]
            [clojure.string :as str]))

(def ^:private _SRC ["https://fred.stlouisfed.org/" "https://www.ici.org/research"])

(def ^:private _NET
  [{"bucket" "us-equities"   "label" "US equities"   "net"  13.1 "inflow" 23.2 "outflow" 10.1}
   {"bucket" "us-govt-bonds" "label" "US Treasuries" "net" -21.7 "inflow"  0.0 "outflow" 21.7}])

(def ^:private _ROT
  [{"from" "us-govt-bonds" "from_label" "US Treasuries"
    "to"   "us-equities"  "to_label"   "US equities"
    "magnitude" 12.4}])

(def ^:private _REGIME
  {"regime" "risk-on" "risk_net" 27.3 "safe_net" -26.6 "no_trade_notice" true})

;; ── helpers ──────────────────────────────────────────────────────────────────

(defmacro expect-raises
  "Assert that body throws an exception whose message contains `contains-str`."
  [contains-str & body]
  `(let [caught# (try ~@body nil (catch Exception e# e#))]
     (is (some? caught#) "expected an exception, none raised")
     (when (seq ~contains-str)
       (is (str/includes? (ex-message caught#) ~contains-str)
           (str "raised but missing " (pr-str ~contains-str) ": " (ex-message caught#))))))

;; ── tests ────────────────────────────────────────────────────────────────────

(deftest test-netflow-post-pins-invariants
  (let [p (social/draft-netflow-post _NET _SRC)]
    (is (= ":dry-run" (get p ":post/status")))
    (is (true?  (get p ":post/is-mirror")))
    (is (true?  (get p ":post/no-trade-notice")))
    (is (false? (get p ":post/server-held-key")))))

(deftest test-netflow-post-body-has-disclaimer
  (let [p (social/draft-netflow-post _NET _SRC)]
    (is (str/includes? (get p ":post/body") "トレードはしない"))))

(deftest test-rotation-post-ok
  (let [p (social/draft-rotation-post _ROT _SRC)]
    (is (str/includes? (get p ":post/body") "US Treasuries"))
    (is (= ":dry-run" (get p ":post/status")))))

(deftest test-regime-post-states-descriptor
  (let [p (social/draft-regime-post _REGIME _SRC)]
    (is (str/includes? (get p ":post/body") "risk-on"))
    (is (str/includes? (get p ":post/body") "助言ではありません"))))

(deftest test-post-requires-two-sources-g3
  (expect-raises "G3"
    (social/draft-netflow-post _NET ["only-one"])))

(deftest test-post-refuses-denied-source
  (expect-raises "Rider"
    (social/draft-netflow-post _NET ["bloomberg terminal" "x"])))

(deftest test-no-trade-guard-blocks-buy
  (expect-raises "G2"
    (#'shionome.methods.social/guard-no-trade "you should buy equities")))

(deftest test-no-trade-guard-blocks-japanese
  (expect-raises "G2"
    (#'shionome.methods.social/guard-no-trade "目標株価は5000円")))

(deftest test-no-trade-guard-allows-disclaimer
  ;; the disclaimer NAMES the prohibited acts; it must not trip the guard
  (is (nil? (#'shionome.methods.social/guard-no-trade social/DISCLAIMER))))

(deftest test-build-live-refuses-g8
  (expect-raises "G8"
    (social/build-live)))

(deftest test-empty-rotation-post-safe
  (let [p (social/draft-rotation-post [] _SRC)]
    (is (= ":dry-run" (get p ":post/status")))))

;; ── runner ───────────────────────────────────────────────────────────────────

(when (= *file* (System/getProperty "babashka.file"))
  (let [{:keys [fail error]} (clojure.test/run-tests 'shionome.methods.test-social)]
    (System/exit (if (zero? (+ fail error)) 0 1))))
